### PR TITLE
Remove external dependencies

### PR DESCRIPTION
**Pull request by tashrifbillah**

### CODE REVIEW

File: pipeline/preprocessing.py [hunk 1]
```diff
@@ -217,41 +220,40 @@ def ANTS_rigid_body_trans(b0_nii, reference=None):
     return (transformed_file, omat_file)
 
 
-def pre_process(lock, subject, reference_list):
+def pre_process(input_file, reference_list, b0_threshold= None, which_bse= '--avg'):
```
Comment (tashrifbillah) [inline pipeline/preprocessing.py]: The default arguments: `b0_threshold= None, which_bse= '--avg'` are introduced to facilitate future modification.

File: pipeline/preprocessing.py [hunk 2]
```diff
@@ -217,41 +220,40 @@ def ANTS_rigid_body_trans(b0_nii, reference=None):
...
-            #dimensions = get_dimension(b0_nii)
-            #cases_dim.append(dimensions)
-            #split_dim.append(int(dimensions[0]))
-            reference_list.append((b0_nii))
+        # convert NRRD/NHDR to NIFIT as the first step
+        # extract bse.py from just NIFTI later
+        if input_file.endswith(SUFFIX_NRRD) | input_file.endswith(SUFFIX_NHDR):
+            inPrefix= input_file.split('.')[0]
+            nifti_write(input_file)
+            input_file= inPrefix+ '.nii.gz'
```
Comment (tashrifbillah) [inline pipeline/preprocessing.py]: Please note that, `nifti_write()` converted file is always `.nii.gz`. Is that okay for your pipeline?

If not, the following could be used here:

```python
if input_file.endswith(SUFFIX_NIFTI):
         Popen('gunzip {input_file}')
```

File: pipeline/preprocessing.py [hunk 3]
```diff
@@ -43,6 +43,7 @@
 output_mask = []
 
 
+# TODO this function is obsolete, remove in future
 def check_gradient(Nhdr_file):
```
Comment (tashrifbillah) [inline pipeline/preprocessing.py]: With my modification, some functions have become obsolete. We should remove them in future ...

**Comment by tashrifbillah** [comment]
Hi @SenthilCaesar , please install two additional requirements and test my branch to make sure it is seamless with rest of your pipeline.

`conversion` and `pnlNipype` are already part of `/path/to/pnlpipe3/bashrc3`. If you source that, they are available to you automatically.

However, if you have a separate system for running this pipeline:

(i) you have to install https://github.com/pnlbwh/conversion using pip (see requirements.txt)
(ii) on the other hand, I trust you to figure out dependencies of 'just' https://github.com/pnlbwh/pnlNipype/blob/master/scripts/bse.py so you can set up a separate `pnlNipype` of your own and run 'just' `bse.py`.

Once you find the modification seamless, we can merge this PR.

File: pipeline/preprocessing.py (hunk 2 above)
Comment (SenthilCaesar) [reply pipeline/preprocessing.py]: Yes. that should be fine

File: pipeline/preprocessing.py (hunk 3 above)
Comment (SenthilCaesar) [reply pipeline/preprocessing.py]: Sure, its better for release

File: pipeline/preprocessing.py (hunk 1 above)
Comment (SenthilCaesar) [reply pipeline/preprocessing.py]: Thank you Tashrif

**Comment by SenthilCaesar** [comment]
> Hi @SenthilCaesar , please install two additional requirements and test my branch to make sure it is seamless with rest of your pipeline.
> 
> `conversion` and `pnlNipype` are already part of `/path/to/pnlpipe3/bashrc3`. If you source that, they are available to you automatically.
> 
> However, if you have a separate system for running this pipeline:
> 
> (i) you have to install https://github.com/pnlbwh/conversion using pip (see requirements.txt)
> (ii) on the other hand, I trust you to figure out dependencies of 'just' https://github.com/pnlbwh/pnlNipype/blob/master/scripts/bse.py so you can set up a separate `pnlNipype` of your own and run 'just' `bse.py`.
> 
> Once you find the modification seamless, we can merge this PR.

Hi Tashrif, I have tested the pipeline and the modification works fine . I think we can merge to the master branch.

**Comment by tashrifbillah** [comment]
Hi @SenthilCaesar , please update the README with installation and running instruction. To be more specific:

- [ ] Mention where should `Model Architecture` etc. be downloaded on a user computer
- [ ] Since prediction is the expected use of the software than training, provide a sample command that gives DWI brain mask given a DWI or whatever
- [ ] Eventually, we should also provide training sample commands.

The goal is to allow me (or a remote user) run this software by their own. Check each box once done.

**Comment by tashrifbillah** [comment]
Hi @SenthilCaesar , does this software run without GPU support? If it does run on just CPU, what packages should we ask the user to install?
https://github.com/pnlbwh/CNN-Diffusion-MRIBrain-Segmentation/tree/remove-ext-dep#install-prerequisites-for-running-the-pipeline

**Comment by SenthilCaesar** [comment]
> Hi @SenthilCaesar , does this software run without GPU support? If it does run on just CPU, what packages should we ask the user to install?
> https://github.com/pnlbwh/CNN-Diffusion-MRIBrain-Segmentation/tree/remove-ext-dep#install-prerequisites-for-running-the-pipeline

Yes, the software runs on CPU . The user should install pip install tensorflow ( this is CPU version ) / pip install tensorflow-gpu ( this is GPU version )

**Comment by tashrifbillah** [comment]
Did you get a chance to test the software installation and execution on a workstation using just CPU support?
For example, we can install a fresh python3 in /tmp/miniconda3 and then follow your README.md to install and execute it.

**Comment by SenthilCaesar** [comment]
> Did you get a chance to test the software installation and execution on a workstation using just CPU support?
> For example, we can install a fresh python3 in /tmp/miniconda3 and then follow your README.md to install and execute it.

Yes, the pipeline was tested on Sylvain workstation CPU.

**Comment by tashrifbillah** [comment]
Hi @SenthilCaesar , please come back to `remove-ext-dep` branch as follows:

```
cd CNN-Diffusion-MRIBrain-Segmentation
cp src/settings.ini src/settings.senthil # back up your config file
git pull origin
git checkout remove-ext-dep
cp src/settings.senthil src/settings.ini
```

We shall do one more round of testing to find potential runtime errors. In the meantime, I would 
like you to review `docs/README.md#training` section. Once we are done, we shall release an affluent master branch.

**Comment by tashrifbillah** [comment]
* @tashrifbillah 
Tests on current `remove-ext-dep` branch:

* 1. `pipeline/dwi_masking.py` works fine on CPU
* 2. `registration.py`, `preprocess_b0.py`, `preprocess_mask.py`, and `maskfilter.py` help messages print properly on CPU without any TensorFlow warning.
* 3. TODO: run each of the scripts in (ii) individually on CPU and run `train.py` on CPU

**Comment by tashrifbillah** [comment]
@SenthilCaesar , please repeat the above tests on your GPU environment and comment like I did so we can keep track of testing.

**Comment by SenthilCaesar** [comment]
pipeline/dwi_masking.py works fine on GPU
src/train.py work fine on GPU
registration.py, preprocess_b0.py, preprocess_mask.py, and maskfilter.py help messages print properly on CPU without any TensorFlow warning

File: pipeline/dwi_masking.py [hunk 4]
```diff
@@ -291,22 +291,23 @@ def npy_to_nifti(b0_normalized_cases, cases_mask_arr, sub_name, view='default',
...
 
         output_filter_file = subject_name[:len(subject_name) - (len(format) + 1)] + '-' + view + '_FilteredMask.nii.gz'
         output_mask_filtered = path.join(output_dir, output_filter_file)
 
-        print('Cleaning up ', CNN_output_file)
-        # mask_filter = "maskfilter -force " + CNN_output_file + " -scale 2 clean " + output_mask_filtered
-        mask_filter = path.join(path.dirname(__file__),'../src/maskfilter') + f' {CNN_output_file} 2 {output_mask_filtered}'
+        #print('Cleaning up ', CNN_output_file)
+        mask_filter = "maskfilter -force " + CNN_output_file + " -scale 2 clean " + output_mask_filtered
+        #mask_filter = path.join(path.dirname(__file__),'../src/maskfilter') + f' {CNN_output_file} 2 {output_mask_filtered}'
```
Comment (tashrifbillah) [inline pipeline/dwi_masking.py]: @SenthilCaesar , did you accidentally disable my maskfilter?

File: src/settings.ini [hunk 5]
```diff
@@ -1,9 +1,9 @@
 [COMMON]
-save_model_dir = "project/model_dir/"
-log_dir = "project/log_dir"
+save_model_dir = "/rfanfs/pnl-zorro/home/sq566/pycharm/Suheyla/data/comp/training"
```
Comment (tashrifbillah) [inline src/settings.ini]: @SenthilCaesar , please bring back the discrete config file I pushed. Again, we need to be careful about not committing every local changes.

File: pipeline/dwi_masking.py [hunk 6]
```diff
@@ -217,13 +217,13 @@ def normalize(b0_resampled):
...
     imgU16 = img.get_data().astype(np.float32)
-    p = np.percentile(imgU16, 99)
+    p = np.percentile(imgU16, percentile)
     data = imgU16 / p
     data[data > 1] = 1
     data[data < 0] = 0
     image_dwi = nib.Nifti1Image(data, img.affine, img.header)
     nib.save(image_dwi, output_file)
-    return output_file
+    data_n.append(output_file)
```
Comment (tashrifbillah) [inline pipeline/dwi_masking.py]: @SenthilCaesar , what's the point of appending to `data_n`, if you are not returning?

File: pipeline/dwi_masking.py (hunk 4 above)
Comment (SenthilCaesar) [reply pipeline/dwi_masking.py]: @tashrifbillah . Nope, mrtrix maskfilter output was little better compared to the python maskfilter . I discussed this with Suheyla. We might have to meet with her and finalize which one to use.

Comment (tashrifbillah) [reply pipeline/dwi_masking.py]: For future changes like this, put a comment so I know what you guys are doing.

File: pipeline/dwi_masking.py (hunk 6 above)
Comment (SenthilCaesar) [reply pipeline/dwi_masking.py]: data_n is a Manager() object . Below resource might be helpful

https://stackoverflow.com/questions/9436757/how-does-multiprocessing-manager-work-in-python

File: pipeline/dwi_masking.py (hunk 4 above)
Comment (tashrifbillah) [reply pipeline/dwi_masking.py]: Moreover, if you haven't finalized anything, let's not make changes.

**Pull request merged**